acvm/src/pwg/blackbox.rs

```diff
@@ -39,56 +45,36 @@ pub(crate) fn solve(
-        }
-        BlackBoxFuncCall { name: BlackBoxFunc::RANGE, inputs, outputs } => {
-            backend.range(initial_witness, inputs, outputs)
+    if !contains_all_inputs(initial_witness, &bb_func.inputs) {
+        if let Some(unassigned_witness) = first_missing_assignment(initial_witness, &bb_func.inputs)
+        {
+            return Ok(OpcodeResolution::Stalled(OpcodeNotSolvable::MissingAssignment(
+                unassigned_witness.0,
+            )));
         }
-        BlackBoxFuncCall { name: BlackBoxFunc::SHA256, inputs, outputs } => {
-            backend.sha256(initial_witness, inputs, outputs)
-        }
-        BlackBoxFuncCall { name: BlackBoxFunc::Blake2s, inputs, outputs } => {
-            backend.blake2s(initial_witness, inputs, outputs)
-        }
-        BlackBoxFuncCall { name: BlackBoxFunc::ComputeMerkleRoot, inputs, outputs } => {
-            backend.compute_merkle_root(initial_witness, inputs, outputs)
-        }
-        BlackBoxFuncCall { name: BlackBoxFunc::SchnorrVerify, inputs, outputs } => {
-            backend.schnorr_verify(initial_witness, inputs, outputs)
-        }
-        BlackBoxFuncCall { name: BlackBoxFunc::Pedersen, inputs, outputs } => {
-            backend.pedersen(initial_witness, inputs, outputs)
+    }
+    match bb_func.name {
+        BlackBoxFunc::AND | BlackBoxFunc::XOR => solve_logic_opcode(initial_witness, bb_func),
+        BlackBoxFunc::RANGE => solve_range_opcode(initial_witness, bb_func),
+        BlackBoxFunc::SHA256 => sha256(initial_witness, bb_func),
+        BlackBoxFunc::Blake2s => blake2s(initial_witness, bb_func),
+        BlackBoxFunc::Keccak256 => keccak256(initial_witness, bb_func),
+        BlackBoxFunc::ComputeMerkleRoot => {
+            backend.compute_merkle_root(initial_witness, &bb_func.inputs, &bb_func.outputs)
         }
-        BlackBoxFuncCall { name: BlackBoxFunc::HashToField128Security, inputs, outputs } => {
-            backend.hash_to_field128_security(initial_witness, inputs, outputs)
+        BlackBoxFunc::SchnorrVerify => {
+            backend.schnorr_verify(initial_witness, &bb_func.inputs, &bb_func.outputs)
         }
-        BlackBoxFuncCall { name: BlackBoxFunc::EcdsaSecp256k1, inputs, outputs } => {
-            backend.ecdsa_secp256k1(initial_witness, inputs, outputs)
+        BlackBoxFunc::Pedersen => {
+            backend.pedersen(initial_witness, &bb_func.inputs, &bb_func.outputs)
         }
-        BlackBoxFuncCall { name: BlackBoxFunc::FixedBaseScalarMul, inputs, outputs } => {
-            backend.fixed_base_scalar_mul(initial_witness, inputs, outputs)
+        BlackBoxFunc::HashToField128Security => {
+            hash_to_field_128_security(initial_witness, bb_func)
         }
-        BlackBoxFuncCall { name: BlackBoxFunc::Keccak256, inputs, outputs } => {
-            backend.keccak256(initial_witness, inputs, outputs)
+        BlackBoxFunc::EcdsaSecp256k1 => secp256k1_prehashed(initial_witness, bb_func),
```

Comment:
This should live next to the other ACVM implemented functions. I'd like to keep these separated from each other.

Reply:
I am not sure what you mean, could you clarify your comment?

Reply:
Done